net/peer_tracker.go

```diff
@@ -18,32 +22,98 @@ var logPeerTracker = logging.Logger("peer-tracker")
 type PeerTracker struct {
 	// mu protects peers
 	mu sync.RWMutex
-	// peers maps stringified peer.IDs to info about their chains
-	peers map[string]*types.ChainInfo
+	// peers maps peer.IDs to info about their chains
+	peers    map[peer.ID]*types.ChainInfo
+	trusted  map[peer.ID]bool
+	updateFn updatePeerFn
 }
+type updatePeerFn func(ctx context.Context, p peer.ID) (*types.ChainInfo, error)
 
 // NewPeerTracker creates a peer tracker.
-func NewPeerTracker() *PeerTracker {
+func NewPeerTracker(trust ...peer.ID) *PeerTracker {
+	trustedSet := make(map[peer.ID]bool)
+	for _, t := range trust {
+		trustedSet[t] = true
+	}
 	return &PeerTracker{
-		peers: make(map[string]*types.ChainInfo),
+		peers:   make(map[peer.ID]*types.ChainInfo),
+		trusted: trustedSet,
+	}
+}
+
+// SetUpdateFn sets the update function `f` on the peer tracker. This function is a prerequisite
+// to the UpdateTrusted logic.
+func (tracker *PeerTracker) SetUpdateFn(f updatePeerFn) {
```

Comment:
This is pretty clunky. It exists because the `PeerTracker` is instantiated when the node is built, the hello protocol in instantiated when the node is started, and the `SayHello` method is used as the callback.